tools/enrollment_preprocessor.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+
+import csv
+import sys
+from argparse import ArgumentParser
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import datetime
+from io import BytesIO
+from pathlib import Path
+from typing import Iterator, NamedTuple, TextIO
+
+from openpyxl import Workbook, load_workbook
+from openpyxl.cell import Cell
+
+
+@dataclass
+class User:
+    title: str
+    last_name: str
+    first_name: str
+    email: str
+
+    def full_name(self) -> str:
+        prefix = ""
+        if self.title:
+            prefix = f"{self.title} "
+        return f"{prefix}{self.first_name} {self.last_name}, {self.email}"
+
+
+class UserCells(NamedTuple):
+    title: Cell | None
+    last_name: Cell
+    first_name: Cell
+    email: Cell
+
+    def _clean(self) -> Iterator[str]:
+        for field in iter(self):
+            if not field or not field.value:
+                yield ""
+                continue
+            field.value = str(field.value or "").strip()
+            yield field.value
+
+    def clean_user(self):
+        return User(*self._clean())
+
+
+def make_bold(text: str) -> str:
+    return f"\033[1m{text}\033[0m"
+
+
+def group_conflicts(
+    users: dict[str, User], user_cells: dict[str, list[UserCells]]
+) -> dict[str, list[tuple[UserCells, str]]]:
+    groups = defaultdict(list)
+    for user_entries in user_cells.values():
+        for cells in user_entries:
+            imported = cells.clean_user()
+
+            if not imported.email:
+                continue
+
+            existing = users.setdefault(imported.email, imported)
+
+            if imported == existing:
+                continue
+
+            for field in ["title", "last_name", "first_name"]:
+                field_value = getattr(cells, field)
+                if field_value is not None and getattr(existing, field) != getattr(imported, field):
+                    groups[field].append((cells, existing.email))
+    return groups
+
+
+def parse_existing(user_data: TextIO) -> dict[str, User]:
+    users = {}
+    reader = csv.reader(user_data, delimiter=";", lineterminator="\n")
+    next(reader)  # skip header
+    for row in reader:
+        user = User(*row)
+        users[user.email] = user
+    return users
+
+
+def parse_imported(enrollment_data: Workbook):
```

Comment:
maybe `user_cells_by_email` (possibly with `get_` or `parse_` prefix)?

Also: Annotate return type

Reply:
Does it make sense to keep the "cells" for such a long time, or could we very early just turn them into a `User` early on? Then, we wouldn't need `UserCells` (which I find a bit unintuitive anyway), and we'd just have the users from the csv file (maybe call them database_users or whatever) and the users from the xls import file (call them import users or whatever).

Reply:
Yes and no. My idea before https://github.com/e-valuation/EvaP/pull/2011/commits/46481147f456a6a7f267f7e08a2e37880c0bef4e was to hold a mapping between user (email) and cells, so that they can later be easily updated with the changed userdata.
Current (46481147f456a6a7f267f7e08a2e37880c0bef4e) parses the excel file again and writes thereby the data.